File: remoter/src/main/java/remoter/compiler/builder/ListParamBuilder.java
package remoter.compiler.builder;

import com.squareup.javapoet.MethodSpec;

import java.util.ArrayList;

import javax.annotation.processing.Messager;
import javax.lang.model.element.Element;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;

/**
 * A {@link ParamBuilder} for List type parameters
 */
class ListParamBuilder extends ParamBuilder {


    protected ListParamBuilder(Messager messager, Element element) {
        super(messager, element);
    }


    @Override
    public void writeParamsToProxy(VariableElement param, ParamType paramType, MethodSpec.Builder methodBuilder) {
        if (param.asType().getKind() == TypeKind.ARRAY) {
            logError("List[] is not supported");
        } else {
            if (paramType != ParamType.OUT) {
                if (isListOfStrings(param.asType())) {
                    methodBuilder.addStatement("data.writeStringList(" + param.getSimpleName() + ")");
                } else {
                    methodBuilder.addStatement("data.writeList(" + param.getSimpleName() + ")");
                }
            }
        }
    }

    @Override
    public void readResultsFromStub(TypeMirror resultType, MethodSpec.Builder methodBuilder) {
        if (resultType.getKind() == TypeKind.ARRAY) {
            logError("List[] is not supported");
        } else {
            if (isListOfStrings(resultType)) {
                methodBuilder.addStatement("reply.writeStringList(result)");
            } else {
                methodBuilder.addStatement("reply.writeList(result)");
            }
        }
    }

    @Override
    public void readOutResultsFromStub(VariableElement param, ParamType paramType, String paramName, MethodSpec.Builder methodBuilder) {
        if (isListOfStrings(param.asType())) {
            methodBuilder.addStatement("reply.writeStringList(" + paramName + ")");
        } else {
            methodBuilder.addStatement("reply.writeList(" + paramName + ")");
        }
    }


    @Override
    public void readResultsFromProxy(TypeMirror resultType, MethodSpec.Builder methodBuilder) {
        if (resultType.getKind() == TypeKind.ARRAY) {
            logError("List[] is not supported");
        } else {
            if (isListOfStrings(resultType)) {
                methodBuilder.addStatement("result = reply.createStringArrayList()");
            } else {
                methodBuilder.addStatement("result = reply.readArrayList(getClass().getClassLoader())");
            }
        }
    }

    @Override
    public void writeParamsToStub(VariableElement param, ParamType paramType, String paramName, MethodSpec.Builder methodBuilder) {
        super.writeParamsToStub(param, paramType, paramName, methodBuilder);
        if (param.asType().getKind() == TypeKind.ARRAY) {
            logError("List[] is not supported");
        } else {
            if (paramType == ParamType.OUT) {
                methodBuilder.addStatement(paramName + " = new $T()", ArrayList.class);
            } else {
                if (isListOfStrings(param.asType())) {
                    methodBuilder.addStatement(paramName + " = data.createStringArrayList()");
                } else {
                    methodBuilder.addStatement(paramName + " = data.readArrayList(getClass().getClassLoader())");
                }
            }
        }
    }

    @Override
    public void readOutParamsFromProxy(VariableElement param, ParamType paramType, MethodSpec.Builder methodBuilder) {
        if (paramType != ParamType.IN) {
            methodBuilder.beginControlFlow("if (" + param.getSimpleName()  + " != null)");
            if (isListOfStrings(param.asType())) {
                methodBuilder.addStatement("reply.readStringList(" + param.getSimpleName() + ")");
            } else {
                methodBuilder.addStatement("reply.readList(" + param.getSimpleName() + ", getClass().getClassLoader())");
            }
            methodBuilder.endControlFlow();
        }
    }

    private boolean isListOfStrings(TypeMirror typeMirror) {
        return typeMirror.toString().equals("java.util.List<java.lang.String>");
    }

}

File: remoter/src/main/java/remoter/compiler/kbuilder/ListParamBuilder.kt
package remoter.compiler.kbuilder

import com.squareup.kotlinpoet.FunSpec
import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.VariableElement
import javax.lang.model.type.TypeKind
import javax.lang.model.type.TypeMirror


/**
 * A [ParamBuilder] for List type parameters
 */
internal class ListParamBuilder(remoterInterfaceElement: Element, bindingManager: KBindingManager) : ParamBuilder(remoterInterfaceElement, bindingManager) {
    override fun writeParamsToProxy(param: VariableElement, paramType: ParamType, methodBuilder: FunSpec.Builder) {
        if (param.asType().kind == TypeKind.ARRAY) {
            logError("List[] is not supported")
        } else {
            if (paramType != ParamType.OUT) {
                if (isListOfStrings(param.asType())) {
                    methodBuilder.addStatement("$DATA.writeStringList(" + param.simpleName + ")")
                } else {
                    methodBuilder.addStatement("$DATA.writeList(" + param.simpleName + ")")
                }
            }
        }
    }

    override fun readResultsFromStub(methodElement: ExecutableElement, resultType: TypeMirror, methodBuilder: FunSpec.Builder) {
        if (resultType.kind == TypeKind.ARRAY) {
            logError("List[] is not supported")
        } else {
            if (isListOfStrings(resultType)) {
                methodBuilder.addStatement("$REPLY.writeStringList($RESULT)")
            } else {
                methodBuilder.addStatement("$REPLY.writeList($RESULT)")
            }
        }
    }

    override fun readOutResultsFromStub(param: VariableElement, paramType: ParamType, paramName: String, methodBuilder: FunSpec.Builder) {
        if (isListOfStrings(param.asType())) {
            methodBuilder.addStatement("$REPLY.writeStringList($paramName)")
        } else {
            methodBuilder.addStatement("$REPLY.writeList($paramName)")
        }
    }

    override fun readResultsFromProxy(methodType: ExecutableElement, methodBuilder: FunSpec.Builder) {
        val resultType = methodType.getReturnAsKotlinType()
        val resultMirror = methodType.getReturnAsTypeMirror()
        if (resultMirror.kind == TypeKind.ARRAY) {
            logError("List[] is not supported")
        } else {
            if (isListOfStrings(resultMirror)) {
                methodBuilder.addStatement("$RESULT = $REPLY.createStringArrayList() as %T ", resultType)
            } else {
                methodBuilder.addStatement("$RESULT = $REPLY.readArrayList(javaClass.getClassLoader()) as %T", resultType)
            }
        }
    }

    override fun writeParamsToStub(methodType: ExecutableElement, param: VariableElement, paramType: ParamType, paramName: String, methodBuilder: FunSpec.Builder) {
        super.writeParamsToStub(methodType, param, paramType, paramName, methodBuilder)
        if (param.asType().kind == TypeKind.ARRAY) {
            logError("List[] is not supported")
        } else {
            if (paramType == ParamType.OUT) {
                methodBuilder.addStatement("$paramName = mutableListOf()")
            } else {
                val suffix = if(param.isNullable()) "" else "!!"
                if (isListOfStrings(param.asType())) {
                    methodBuilder.addStatement("$paramName = $DATA.createStringArrayList()$suffix")
                } else {
                    methodBuilder.addStatement("$paramName = $DATA.readArrayList(getClass().getClassLoader())$suffix")
                }
            }
        }
    }

    override fun readOutParamsFromProxy(param: VariableElement, paramType: ParamType, methodBuilder: FunSpec.Builder) {
        if (paramType != ParamType.IN) {
            if (param.isNullable()){
                methodBuilder.beginControlFlow("if (${param.simpleName} != null)")
            }

            if (isListOfStrings(param.asType())) {
                methodBuilder.addStatement("$REPLY.readStringList(" + param.simpleName + ")")
            } else {
                methodBuilder.addStatement("$REPLY.readList(" + param.simpleName + ", getClass().getClassLoader())")
            }

            if (param.isNullable()){
                methodBuilder.endControlFlow()
            }
        }
    }

    private fun isListOfStrings(typeMirror: TypeMirror): Boolean {
        return typeMirror.toString() == "java.util.List<java.lang.String>"
    }
}
